helmfile.d/charts/prometheus-alerts/files/cluster-autoscaler.yaml

```diff
@@ -0,0 +1,142 @@
+      labels:
+        severity: warning
+      annotations:
+        summary: "A node group is in backoff state"
+        description: "The node group {{"{{ $labels.node_group }}"}} in cluster {{"`{{ $labels.cluster }}`"}} is currently in a backoff state due to reason: {{"{{ $labels.reason }}"}}. This means that the Cluster Autoscaler will not try to scale it up."
+        runbook_url: {{ .Values.runbookUrls.clusterAutoscaler.ClusterAutoscalerNodeGroupBackoff }}
+
+    - alert: ClusterAutoscalerUnneededNodesNotScaledDown
+      expr: cluster_autoscaler_unneeded_nodes_count > 0
+      for: 24h
+      labels:
+        severity: warning
+      annotations:
+        summary: "Cluster Autoscaler reports unneeded nodes that are not being scaled down"
+        description: "There are {{"{{ $value }}"}} nodes in cluster {{"`{{ $labels.cluster }}`"}} identified as unneeded by the Cluster Autoscaler, but they have not been scaled down for 24 hours."
+
+    # TODO: The reason label is just the reason constant integer, needs to be fixed upstream. Perhaps a new label like "reason_str" or propose to change it to the name of the constant.
+    # TODO: Doesn't tell you which node it is. Needs to be fixed upstream.
+    # - alert: ClusterAutoscalerUnremovableNodesStuck
+    #   expr: cluster_autoscaler_unremovable_nodes_count > 0
+    #   for: 15m
+    #   labels:
+    #     severity: warning
+    #   annotations:
+    #     summary: "Cluster Autoscaler reports unremovable nodes"
+    #     description: "There are {{"{{ $value }}"}} node(s) that the Cluster Autoscaler cannot remove due to reason: {{"{{ $labels.reason }}"}}."
+    #     runbook_url: {{ .Values.runbookUrls.clusterAutoscaler.ClusterAutoscalerUnremovableNodesStuck }}
+
+    - alert: ClusterAutoscalerClusterAutoscalerErrors
+      expr: sum(rate(cluster_autoscaler_errors_total[5m])) by (cluster, type) > 0
+      for: 1m
+      labels:
+        severity: high
+      annotations:
+        summary: "Cluster Autoscaler is reporting errors"
+        description: "The Cluster Autoscaler for cluster {{"`{{ $labels.cluster }}`"}} is experiencing errors of type {{"{{ $labels.type }}"}}."
+        runbook_url: {{ .Values.runbookUrls.clusterAutoscaler.ClusterAutoscalerClusterAutoscalerErrors }}
+
+    - alert: ClusterAutoscalerPendingNodeDeletionsHigh
+      expr: cluster_autoscaler_pending_node_deletions > 0
+      for: 30m
+      labels:
+        severity: warning
+      annotations:
+        summary: "Nodes are stuck in pending deletion state"
+        description: "There are {{"{{ $value }}"}} nodes in cluster {{"`{{ $labels.cluster }}`"}} that have completed scale-down but are still pending deletion."
+        runbook_url: {{ .Values.runbookUrls.clusterAutoscaler.ClusterAutoscalerPendingNodeDeletionsHigh }}
+
+    - alert: ClusterAutoscalerScaleUpTooLong
+      expr: histogram_quantile(0.99, sum by (le, cluster) (rate(cluster_autoscaler_function_duration_seconds_bucket{function="scaleUp"}[5m]))) > 60
```

Comment:
Where did you get the 60 seconds from? Is it from looking at real world data or from some documentation?

Reply:
I don't have any more specific reason, but I think that 60 seconds seems reasonable.

Reply:
Just checked some empirical data. And seems to be ~7-9 seconds in normal cases, ~20 seconds spikes is not uncommon (maybe it's when it's doing stuff). And I saw one spike of 700 seconds, but it seems to have been some issue or so at that time so that makes sense.

And with that I also agree that 60 seconds seems like a good value

Reply:
Great, thanks for looking at some real-world data. Much more representative compared to my dev cluster stats. :smile: